pallets/offences/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+		// were spotted in unique reports.
+		let TriageOutcome { concurrent_offenders } =
+			match Self::triage_offence_report::<O>(reporters, &time_slot, offenders) {
+				Some(triage) => triage,
+				None => return Err(OffenceError::DuplicateReport),
+			};
+
+		// Define the slash strategy.
+		let slash_strategy = if O::ID == *b"im-online:offlin" {
+			SlashStrategy::Disconnect
+		} else {
+			SlashStrategy::BlackList
+		};
+
+		T::OnOffenceHandler::on_offence(
+			&concurrent_offenders,
+			slash_strategy,
+			offence.session_index(),
+		);
+
+		Self::deposit_event(Event::Offence { kind: O::ID, timeslot: time_slot.encode() });
+
+		Ok(())
+	}
+
+	fn is_known_offence(offenders: &[T::IdentificationTuple], time_slot: &O::TimeSlot) -> bool {
+		let any_unknown = offenders.iter().any(|offender| {
+			let report_id = Self::report_id::<O>(time_slot, offender);
+			!<Reports<T>>::contains_key(report_id)
+		});
+
+		!any_unknown
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Compute the ID for the given report properties.
+	///
+	/// The report id depends on the offence kind, time slot and the id of
+	/// offender.
+	fn report_id<O: Offence<T::IdentificationTuple>>(
+		time_slot: &O::TimeSlot,
+		offender: &T::IdentificationTuple,
+	) -> ReportIdOf<T> {
+		(O::ID, time_slot.encode(), offender).using_encoded(T::Hashing::hash)
+	}
+
+	/// Triages the offence report and returns the set of offenders that was
+	/// involved in unique reports along with the list of the concurrent
+	/// offences.
+	fn triage_offence_report<O: Offence<T::IdentificationTuple>>(
+		reporters: Vec<T::AccountId>,
+		time_slot: &O::TimeSlot,
+		offenders: Vec<T::IdentificationTuple>,
+	) -> Option<TriageOutcome<T>> {
+		let mut storage = ReportIndexStorage::<T, O>::load(time_slot);
+
+		let mut any_new = false;
+		for offender in offenders {
+			let report_id = Self::report_id::<O>(time_slot, &offender);
+
+			if !<Reports<T>>::contains_key(report_id) {
+				any_new = true;
+				<Reports<T>>::insert(
+					report_id,
+					OffenceDetails { offender, reporters: reporters.clone() },
+				);
+
+				storage.insert(report_id);
+			}
+		}
+
+		if any_new {
+			// Load report details for the all reports happened at the same time.
+			let concurrent_offenders = storage
+				.concurrent_reports
+				.iter()
+				.filter_map(<Reports<T>>::get)
+				.collect::<Vec<_>>();
+
+			storage.save();
+
+			Some(TriageOutcome { concurrent_offenders })
+		} else {
+			None
+		}
+	}
+}
+
+struct TriageOutcome<T: Config> {
+	/// Other reports for the same report kinds.
+	concurrent_offenders: Vec<OffenceDetails<T::AccountId, T::IdentificationTuple>>,
+}
+
+/// An auxiliary struct for working with storage of indexes localized for a
+/// specific offence kind (specified by the `O` type parameter).
+///
+/// This struct is responsible for aggregating storage writes and the underlying
+/// storage should not accessed directly meanwhile.
+#[must_use = "The changes are not saved without called `save`"]
```

Comment:
Question: What does the `#[must_use = "something"]` macro do?

Reply:
The must_use attribute is used to issue a diagnostic warning when a value is not "used". https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute